ethereum/core/src/main/java/org/hyperledger/besu/ethereum/trie/diffbased/common/trielog/TrieLogPruner.java

```diff
@@ -83,38 +90,82 @@ public TrieLogPruner(
             BesuMetricCategory.PRUNER, "trie_log_pruned_orphan", "trie log pruned orphan");
   }
 
-  public int initialize() {
-    return preloadQueue();
+  public void initialize() {
+    preloadQueueWithTimeout();
   }
 
-  private int preloadQueue() {
+  private void preloadQueueWithTimeout() {
+
     LOG.atInfo()
-        .setMessage("Loading first {} trie logs from database...")
+        .setMessage("Attempting to load first {} trie logs from database...")
         .addArgument(loadingLimit)
         .log();
+
+    try (final ScheduledExecutorService preloadExecutor = Executors.newScheduledThreadPool(1)) {
+
+      final AtomicBoolean timeoutOccurred = new AtomicBoolean(false);
+      final Runnable timeoutTask =
+          () -> {
+            timeoutOccurred.set(true);
+            LOG.atWarn()
+                .setMessage(
+                    "Timeout occurred while loading and processing {} trie logs from database")
+                .addArgument(loadingLimit)
+                .log();
+          };
+
+      final ScheduledFuture<?> timeoutFuture =
+          preloadExecutor.schedule(timeoutTask, PRELOAD_TIMEOUT_IN_SECONDS, TimeUnit.SECONDS);
+      LOG.atInfo()
+          .setMessage(
+              "Trie log pruning will timeout after {} seconds. If this is timing out, consider using `besu storage trie-log prune` subcommand, see https://besu.hyperledger.org/public-networks/how-to/bonsai-limit-trie-logs")
+          .addArgument(PRELOAD_TIMEOUT_IN_SECONDS)
+          .log();
+
+      preloadQueue(timeoutOccurred, timeoutFuture);
+    }
+  }
+
+  private void preloadQueue(
+      final AtomicBoolean timeoutOccurred, final ScheduledFuture<?> timeoutFuture) {
+
     try (final Stream<byte[]> trieLogKeys = rootWorldStateStorage.streamTrieLogKeys(loadingLimit)) {
```

Comment:
Would it make sense to start streaming at a random key? There is a method streamFromKey() in class KeyValueStorage that allows you to pass in the starting key.